jetbrains-core/src/software/aws/toolkits/jetbrains/ui/feedback/SubmitFeedbackPanel.kt

```diff
@@ -0,0 +1,108 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.ui.feedback
+
+import com.intellij.ui.components.JBLabel
+import com.intellij.ui.components.JBRadioButton
+import com.intellij.ui.components.JBScrollPane
+import com.intellij.ui.components.JBTextArea
+import com.intellij.util.IconUtil
+import com.intellij.util.text.nullize
+import icons.AwsIcons
+import org.jetbrains.annotations.TestOnly
+import software.amazon.awssdk.services.toolkittelemetry.model.Sentiment
+import software.aws.toolkits.jetbrains.services.telemetry.ClientMetadata
+import software.aws.toolkits.resources.message
+import java.net.URLEncoder
+import javax.swing.ButtonGroup
+import javax.swing.JLabel
+import javax.swing.JPanel
+
+class SubmitFeedbackPanel(initiallyPositive: Boolean = true) {
+    private lateinit var rootPanel: JPanel
+    private lateinit var smileButton: JBRadioButton
+    private lateinit var sadButton: JBRadioButton
+    private lateinit var smileIcon: JLabel
+    private lateinit var sadIcon: JLabel
+    private lateinit var textArea: JBTextArea
+    private lateinit var textAreaPane: JBScrollPane
+    private lateinit var lengthLimitLabel: JLabel
+    private lateinit var githubLink: JBLabel
+    private lateinit var sentimentButtonGroup: ButtonGroup
+
+    val panel: JPanel
+        get() = rootPanel
+
+    val sentiment: Sentiment?
+        get() = when {
+            smileButton.isSelected -> Sentiment.POSITIVE
+            sadButton.isSelected -> Sentiment.NEGATIVE
+            else -> null
+        }
+
+    var comment: String?
+        get() = textArea.text?.nullize(true)
+        @TestOnly
+        internal set(value) { textArea.text = value }
+
+    private fun createUIComponents() {
+        textArea = JBTextArea(6, 70)
+        textArea.lineWrap = true
+        textAreaPane = JBScrollPane(textArea)
+        smileIcon = JLabel(IconUtil.scale(AwsIcons.Misc.SMILE, null, 3f))
+        sadIcon = JLabel(IconUtil.scale(AwsIcons.Misc.FROWN, null, 3f))
+        lengthLimitLabel = JLabel()
+
+        textArea.document.addUndoableEditListener { onTextAreaUpdate() }
+    }
+
+    init {
+        // runs after $$$setupUI$$$
+        // null out placeholder text
+        smileIcon.text = null
+        sadIcon.text = null
+
+        // select initial value
+        if (initiallyPositive) {
+            smileButton.isSelected = true
+        } else {
+            sadButton.isSelected = true
+        }
+
+        // update remaining character count
+        onTextAreaUpdate()
+        // make links work
+        githubLink.setCopyable(true)
+    }
+
+    private fun onTextAreaUpdate() {
+        val currentLength = comment?.length ?: 0
+        lengthLimitLabel.text = message("feedback.limit.label", currentLength, MAX_LENGTH)
+        if (currentLength >= MAX_LENGTH) {
+            lengthLimitLabel.text = message("feedback.limit.reached", MAX_LENGTH)
```

Comment:
The common pattern I've seen is that the label turns red to indicated that the limit is exceeded rather than changing the text itself.

So it says like `2005/2000` in red text

Reply:
Also, looks like VSCode feedback just prevents you from typing once you reach the maximum. That seems like another good option.

Reply:
went with red text